pooch/downloaders.py

```diff
@@ -308,3 +309,118 @@ def callback(data):
+        requires authentication (i.e., no anonymous SFTP). Use the empty string
+        to indicate no password is required.
+    timeout : int
+        Timeout in seconds for sftp socket operations, use None to mean no
+        timeout.
+    progressbar : bool
+        If True, will print a progress bar of the download to standard error
+        (stderr). Requires `tqdm <https://github.com/tqdm/tqdm>`__ to be
+        installed.
+
+    """
+
+    def __init__(
+        self,
+        port=22,
+        username="anonymous",
+        password="",
+        account="",
+        timeout=None,
+        progressbar=False,
+    ):
+
+        self.port = port
+        self.username = username
+        self.password = password
+        self.account = account
+        self.timeout = timeout
+        self.progressbar = progressbar
+        if self.progressbar and tqdm is None:
+            raise ValueError("Missing package 'tqdm' required for progress bars.")
+
+    def __call__(self, url, output_file, pooch):
+        """
+        Download the given URL over SFTP to the given output file.
+
+        Parameters
+        ----------
+        url : str
+            The URL to the file you want to download.
+        output_file : str or file-like object
+            Path (and file name) to which the file will be downloaded.
+        pooch : :class:`~pooch.Pooch`
+            The instance of :class:`~pooch.Pooch` that is calling this method.
+        """
+
+        parsed_url = parse_url(url)
+        ispath = not hasattr(output_file, "write")
+        if ispath:
+            output_obj = open(output_file, "wb")
+        try:
+            cnopts = pysftp.CnOpts()
+            cnopts.hostkeys = None
+            sftp = pysftp.Connection(
+                host=parsed_url["netloc"],
+                port=self.port,
+                username=self.username,
+                password=self.password,
+                cnopts=cnopts,
+            )
+            sftp.timeout = self.timeout
+
+            if self.progressbar:
+
+                class TqdmWrap(tqdm):
+                    """Wrapper for tqdm to show sftp progress"""
+
+                    def view_bar(self, progr, size):
+                        """callback for sftp.get"""
+                        self.total = int(size)
+                        self.update(int(progr - self.n))
+
+                with output_obj:
+                    size = int(sftp.stat(parsed_url["path"]).st_size)
+                    use_ascii = bool(sys.platform == "win32")
+                    progress = TqdmWrap(
+                        total=size,
+                        ncols=79,
+                        ascii=use_ascii,
+                        unit="B",
+                        unit_scale=True,
+                        leave=True,
+                    )
+
+                    with progress:
+                        sftp.get(
+                            parsed_url["path"], output_file, callback=progress.view_bar
```

Comment:
Could this not be a callback function instead of a class like in the FTP version?